File: python/src/convomem/resources/entities.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any

from convomem._types import Entity, EntityGraphResponse, EntityListResponse
from convomem.resources._base import _BaseResource

if TYPE_CHECKING:
    from convomem._async_client import AsyncConvoMemClient
    from convomem._client import ConvoMemClient
    from convomem._types import EntitySearchParams


class EntitiesResource(_BaseResource):
    """Synchronous resource for managing knowledge graph entities."""

    def __init__(self, client: ConvoMemClient) -> None:
        self._client = client

    def list(
        self,
        *,
        page: int | None = None,
        limit: int | None = None,
        type: str | None = None,
    ) -> Any:
        """List entities with pagination and optional type filter."""
        params: dict[str, str] = {}
        if page is not None:
            params["page"] = str(page)
        if limit is not None:
            params["limit"] = str(limit)
        if type is not None:
            params["type"] = type
        return self._client._request(
            "GET", "/entities", params=params, response_type=EntityListResponse
        )

    def get(
        self,
        entity_id: str,
    ) -> Any:
        """Get a single entity by its unique ID."""
        return self._client._request("GET", f"/entities/{entity_id}", response_type=Entity)

    def search(
        self,
        params: EntitySearchParams,
    ) -> Any:
        """Search entities by query text with optional type filter."""
        query: dict[str, str] = {"query": params.query}
        if params.type is not None:
            query["type"] = params.type
        if params.limit is not None:
            query["limit"] = str(params.limit)
        return self._client._request(
            "GET", "/entities/search", params=query, response_type=EntityListResponse
        )

    def get_graph(
        self,
        *,
        entity_id: str | None = None,
        depth: int | None = None,
    ) -> Any:
        """Get the entity relationship graph."""
        params: dict[str, str] = {}
        if entity_id is not None:
            params["entityId"] = entity_id
        if depth is not None:
            params["depth"] = str(depth)
        return self._client._request(
            "GET", "/entities/graph", params=params, response_type=EntityGraphResponse
        )

    def delete(
        self,
        entity_id: str,
    ) -> None:
        """Delete an entity permanently."""
        self._client._request("DELETE", f"/entities/{entity_id}")


class AsyncEntitiesResource(_BaseResource):
    """Asynchronous resource for managing knowledge graph entities."""

    def __init__(self, client: AsyncConvoMemClient) -> None:
        self._client = client

    async def list(
        self,
        *,
        page: int | None = None,
        limit: int | None = None,
        type: str | None = None,
    ) -> Any:
        """List entities with pagination and optional type filter."""
        params: dict[str, str] = {}
        if page is not None:
            params["page"] = str(page)
        if limit is not None:
            params["limit"] = str(limit)
        if type is not None:
            params["type"] = type
        return await self._client._request(
            "GET", "/entities", params=params, response_type=EntityListResponse
        )

    async def get(
        self,
        entity_id: str,
    ) -> Any:
        """Get a single entity by its unique ID."""
        return await self._client._request("GET", f"/entities/{entity_id}", response_type=Entity)

    async def search(
        self,
        params: EntitySearchParams,
    ) -> Any:
        """Search entities by query text with optional type filter."""
        query: dict[str, str] = {"query": params.query}
        if params.type is not None:
            query["type"] = params.type
        if params.limit is not None:
            query["limit"] = str(params.limit)
        return await self._client._request(
            "GET", "/entities/search", params=query, response_type=EntityListResponse
        )

    async def get_graph(
        self,
        *,
        entity_id: str | None = None,
        depth: int | None = None,
    ) -> Any:
        """Get the entity relationship graph."""
        params: dict[str, str] = {}
        if entity_id is not None:
            params["entityId"] = entity_id
        if depth is not None:
            params["depth"] = str(depth)
        return await self._client._request(
            "GET", "/entities/graph", params=params, response_type=EntityGraphResponse
        )

    async def delete(
        self,
        entity_id: str,
    ) -> None:
        """Delete an entity permanently."""
        await self._client._request("DELETE", f"/entities/{entity_id}")

File: typescript/resources/entities.ts
import type { ConvoMemClient } from "../client.ts";
import type {
  Entity,
  EntityGraphResponse,
  EntityListResponse,
  EntitySearchParams,
} from "../types.ts";

export class EntitiesResource {
  #client: ConvoMemClient;

  constructor(client: ConvoMemClient) {
    this.#client = client;
  }

  /**
   * List entities (paginated).
   */
  async list(
    opts?: { page?: number; limit?: number; type?: string },
  ): Promise<EntityListResponse> {
    const params: Record<string, string> = {};
    if (opts?.page) params.page = String(opts.page);
    if (opts?.limit) params.limit = String(opts.limit);
    if (opts?.type) params.type = opts.type;

    return await this.#client.request<EntityListResponse>(
      "GET",
      "/entities",
      { params },
    );
  }

  /**
   * Get a single entity by ID.
   */
  async get(entityId: string): Promise<Entity> {
    return await this.#client.request<Entity>(
      "GET",
      `/entities/${entityId}`,
    );
  }

  /**
   * Search entities by query parameters.
   */
  async search(params: EntitySearchParams): Promise<EntityListResponse> {
    const query: Record<string, string> = {};
    if (params.query) query.query = params.query;
    if (params.type) query.type = params.type;
    if (params.limit) query.limit = String(params.limit);

    return await this.#client.request<EntityListResponse>(
      "GET",
      "/entities/search",
      { params: query },
    );
  }

  /**
   * Get entity relationship graph.
   */
  async getGraph(
    opts?: { entityId?: string; depth?: number },
  ): Promise<EntityGraphResponse> {
    const params: Record<string, string> = {};
    if (opts?.entityId) params.entityId = opts.entityId;
    if (opts?.depth) params.depth = String(opts.depth);

    return await this.#client.request<EntityGraphResponse>(
      "GET",
      "/entities/graph",
      { params },
    );
  }

  /**
   * Delete an entity by ID.
   */
  async delete(entityId: string): Promise<void> {
    return await this.#client.request(
      "DELETE",
      `/entities/${entityId}`,
    );
  }
}
